### ISSUE TITLE
Add rayon?

### ISSUE THREAD

**Issue by killercup** [killercup/quicli#4]
Are there any CLI-typical tasks that can be done in parallel? Does exposing rayon's `.par_iter` make more users take advantage of this feature?

https://docs.rs/rayon

**Comment by polarathene** [comment]
I think that the developer is likely to use rayon within their own processing logic, eg I have a tool that currently is single threaded that just reads a binary file and divides it into smaller files for being able to pass onto a third party process which is RAM heavy(about 8x the binary size, eg 8GB tends to use around 64GB RAM, datasets usually exceed 40GB) unable to handle the whole dataset in one go.

Maybe I misunderstand this crate, I am under the impression it allows a dev to add CLI support more easily. When it comes to running subprocesses and passing on some configuration like I do with a third party tool, running single threaded instances in parallel can be useful if the resources allow for it(in my case I may want to run some logic/expression on hardware stats to know what is the optimal amount of instances of the subprocess that can be run/handled as all cores may crash/fail(not enough RAM, not enough data to consume, eg list of files in a directory), having args to override is fine though. 

[parallel](https://crates.io/crates/parallel) might be good for supporting this over rayon? At least my description is more appropriate to it as a "work stealer"/load balancer I think? See the docs for how they've already gotten options for applying constraints like on num cores.

**Comment by killercup** [comment]
> Maybe I misunderstand this crate, I am under the impression it allows a dev to add CLI support more easily.

My goal is to make writing small stand-alone CLI apps easy, and while I won't focus on them, I'd be happy if quicli also fits other use-cases.

If you are using this to add a CLI to an exiting tool, it might not make much sense to use rayon specifically, I agree. You can use rayon to act as a work-stealing queue for spawning subprocess but that is not its primary use-case.

> the developer is likely to use rayon within their own processing logic

Exactly, and if that processing logic is not that complex (as in: can live in the CLI's main function), it'd be quite handy to have some simple tools to speed this up.

**tl;dr** When I wrote this issue, I was thinking of speeding up something like `read_file("foo")?.lines().filter(|line| line.starts_with("BAR")).map(…)` by using `par_lines` instead of `lines`.

> parallel might be good for supporting this over rayon?

Interesting! Parallel is a CLI tool itself, and about running _processes_ in parallel, right? What you describe sound more sophisticated than I feel like we can give a general solution for. I'd rather say "use crate X if your project depends on doing this right" than reinventing/re-exporting a whole library.
